language-support/ts/daml-ledger/index.ts

```diff
@@ -608,44 +628,147 @@ class Ledger {
-    const request = [{templateId: template.templateId, key}];
-    const reconnectRequest = (): object[] => [{...request[0], 'contractIdAtOffset': lastContractId}]
-    const change = (contract: CreateEvent<T, K, I> | null, events: readonly Event<T, K, I>[]): CreateEvent<T, K, I> | null => {
-      // NOTE(MH, #4564): We're very lenient here. We should not see a create
-      // event when `contract` is currently not null. We should also only see
-      // archive events when `contract` is currently not null and the contract
-      // ids match. However, the JSON API does not provied these guarantees yet
-      // but we're working on them.
+    return this.streamFetchByKeyCommon(template, [key], "streamFetchByKey", (cs) => cs[0], (c) => [c]);
+  }
+
+  /**
+   * @internal
+   *
+   * Returns the same API as [[streamSubmit]] but does not, in fact, establish
+   * any socket connection. Instead, this is a stream that always has the given
+   * value as its accumulated state.
+   */
+  constantStream<T extends object, K, I extends string, V>(
+    value: V,
+  ): Stream<T, K, I, V> {
+    function on(t: 'live', l: (v: V) => void): void;
+    function on(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function on(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function on(type: any, listener: any): any {
+        if (type === 'live') {
+          listener(value);
+        }
+        if (type === 'change') {
+          listener(value, []);
+        }
+    }
+    function off(t: 'live', l: (v: V) => void): void;
+    function off(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function off(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+    function off(_t: any, _l: any): any {}
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { on, off, close: (): void => {} };
+  }
+
+  /**
+   * @internal
+   *
+   */
+  streamFetchByKeyCommon<T extends object, K, I extends string, S>(
+    template: Template<T, K, I>,
+    keys: K[], // non-zero length
+    name: string,
+    to: (contracts: (CreateEvent<T, K, I> | null)[]) => S,
+    from: (state: S) => (CreateEvent<T, K, I> | null)[],
+  ): Stream<T, K, I, S> {
+    const lastContractIds: (ContractId<T> | null)[] = Array(keys.length).fill(null);
+    const keysCopy = Array.from(keys);
+    const initState: (CreateEvent<T, K, I> | null)[] = Array(keys.length).fill(null);
+    const request = keys.map(k => ({templateId: template.templateId, key: k}));
+    const reconnectRequest = (): object[] => request.map((r, idx) => ({...r, 'contractIdAtOffset': lastContractIds[idx]}))
+    const change = (state: S, events: readonly Event<T, K, I>[]): S => {
+      const contracts = from(state);
```

Comment:
Perhaps worth mentioning in a comment that this currently does a linear search over the contract keys. We could think about optimizing it but I expect the number of keys to be quite small so I wouldn’t worry about this for now.

Reply:
I had actually started with an optimized version, using a map of key to index. but then I thought back to the dynamic list that justified the 0-length version, and I saw no reason to force users to give us a set of keys, rather than just a list. If they want to pin the same key multiple times, I have no issue giving them the same contract multiple times. That's why there is no escape to the loop.

We could still try to optimize this, say with a map of key to `number[]`, I suppose.

Reply:
Good point, I do agree with giving them the same contract multiple times.